File: Parse/Router/GenApi/Templates/flutter/Lib/http.dart
import 'dart:async';
import 'dart:typed_data';
import 'package:dio/dio.dart';
import 'package:fixnum/fixnum.dart';

import '../config.dart';
import 'crypto.dart';
import 'auth.dart';
import 'cache_mgr.dart';
import 'network.dart';

// 导入 Protobuf 通用响应类型
// 注意：实际使用时需要根据项目调整导入路径
// import 'package:your_app/proto/generated/common.pb.dart';

/// 显示 API 错误弹窗
/// 需要在实际项目中实现具体的 UI 显示逻辑
void showApiError(String content, {String title = '请求错误'}) {
  // TODO: 实现错误弹窗显示
  // 可以使用 Get.snackbar, showDialog 等
  print('[$title] $content');
}

/// 生成缓存键
String getCacheKey(String router, [Map<String, dynamic>? params]) {
  String cacheKey = 'Api:$router';

  if (params == null || params.isEmpty) {
    return cacheKey;
  }

  final paramValues = params.values.map((value) {
    if (value is Map) {
      return value.values.join('');
    }
    return value.toString();
  }).join('');

  return paramValues.isNotEmpty ? '$cacheKey:$paramValues' : cacheKey;
}

/// HTTP 客户端类
class Http {
  /// 单例实例
  static final Http instance = Http._internal();

  /// Dio 实例
  late final Dio _dio;

  /// Token 刷新 Completer（防止并发刷新）
  static Completer<void>? _refreshTokenCompleter;

  Http._internal() {
    _dio = Dio(
      BaseOptions(
        connectTimeout: Duration(milliseconds: Config.timeout),
        receiveTimeout: Duration(milliseconds: Config.timeout),
        responseType: ResponseType.bytes,
      ),
    );
  }

  /// 调用 API
  /// [url] API 完整 URL
  /// [params] 请求参数（Protobuf 对象）
  /// [showError] 是否显示错误
  /// [errorMessage] 错误提示消息
  /// [toProto] 将参数转换为字节的函数
  /// [fromProto] 从字节解析响应的函数
  /// [cacheTime] 缓存时间（秒），0 表示不缓存
  static Future<T> callApi<T>({
    required String url,
    dynamic params,
    bool showError = true,
    required String errorMessage,
    Uint8List Function(dynamic)? toProto,
    required T Function(Uint8List) fromProto,
    int cacheTime = 0,
  }) async {
    final cacheKey = getCacheKey(url, params is Map<String, dynamic> ? params : null);

    // 检查缓存
    if (cacheTime > 0) {
      final cacheData = await CacheMgr.get<Uint8List>(cacheKey);
      if (cacheData != null) {
        try {
          return fromProto(cacheData);
        } catch (e) {
          // 缓存解析失败，继续请求
        }
      }
    }

    // 编码请求参数
    Uint8List? sendParams;
    if (params != null && toProto != null) {
      sendParams = toProto(params);
    }

    // 发送请求
    try {
      final response = await Http.instance._send(url, sendParams);

      // 设置缓存
      if (cacheTime > 0) {
        await CacheMgr.set(cacheKey, response, expired: cacheTime);
      }

      return fromProto(response);
    } catch (e) {
      if (showError) {
        showApiError(e.toString(), title: errorMessage);
      }
      rethrow;
    }
  }

  /// 使用新 Token 重试请求
  Future<Uint8List> _retryRequestWithNewToken(String url, Uint8List? data) async {
    // 如果已经有正在进行的刷新 Token 请求，直接等待它
    if (_refreshTokenCompleter != null) {
      await _refreshTokenCompleter!.future;
    } else {
      _refreshTokenCompleter = Completer<void>();
      try {
        await Auth.refreshToken();
        _refreshTokenCompleter!.complete();
      } catch (e) {
        _refreshTokenCompleter!.completeError(e);
        rethrow;
      } finally {
        _refreshTokenCompleter = null;
      }
    }

    // 使用新 Token 重试原始请求
    return await _send(url, data, allowRetry: false);
  }

  /// 发送 HTTP 请求
  /// [url] 请求 URL
  /// [data] 请求数据（Protobuf 编码后的数据）
  /// [allowRetry] 是否允许 Token 过期后重试
  Future<Uint8List> _send(
    String url,
    Uint8List? data, {
    bool allowRetry = true,
  }) async {
    // 等待网络就绪（处理首次打开权限问题）
    await Network.waitForReady();

    final nowTime = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    final random = DateTime.now().microsecond % 10000;

    // 构建请求头
    final headers = <String, String>{
      'Content-Type': 'application/x-protobuf',
      'time': nowTime.toString(),
      'random': random.toString(),
      'app-id': Config.appId,
      'url': url,
      'token': Crypto.sign(url, random, nowTime, Config.appSecret),
      'x-forwarded-proto': Config.apiUrl.startsWith('https') ? 'https' : 'http',
    };

    // 添加语言设置
    final currentLang = await CacheMgr.get<String>('currentLang');
    headers['lang'] = currentLang ?? Config.defaultLanguage;

    // 添加短 Token
    final shortToken = Auth.getShortToken();
    if (shortToken != null && shortToken.isNotEmpty) {
      headers['authorization'] = shortToken;
    }

    try {
      final response = await _dio.post<List<int>>(
        url,
        data: data,
        options: Options(
          headers: headers,
          responseType: ResponseType.bytes,
        ),
      );

      // 处理 401 未授权
      if (response.statusCode == 401 && allowRetry) {
        return await _retryRequestWithNewToken(url, data);
      }

      // 处理非 200 状态码
      if (response.statusCode != 200) {
        throw Exception('HTTP Error: ${response.statusCode}');
      }

      // 检查响应数据
      if (response.data == null) {
        throw Exception('响应数据为空');
      }

      // 返回响应数据
      // 注意：实际项目中需要解析 Protobuf 通用响应，检查 errno
      return Uint8List.fromList(response.data!);
    } on DioException catch (e) {
      if (e.response?.statusCode == 401 && allowRetry) {
        return await _retryRequestWithNewToken(url, data);
      }
      rethrow;
    }
  }
}

File: Parse/Router/GenApi/Templates/ts/Lib/Http.ts
/**
 * HTTP 客户端
 * 封装 UniApp 的网络请求，支持 Protobuf、缓存、签名、Token 刷新等功能
 */

import { Protobuf } from "@/proto/proto.js";
import { Config } from "../Config";
import { Crypto } from "./Crypto";
import { Auth } from "./Auth";
import { CacheMgr } from "./CacheMgr";
import { Network } from "./Network";

// 声明全局 uni 对象类型（UniApp环境）
declare const uni: any;

/**
 * 显示 API 错误弹窗
 * @param content 错误内容
 * @param title 错误标题
 */
export const showApiError = async (content: string, title: string = "请求错误") => {
    uni.showModal({
        title: title,
        content: content,
        confirmText: '确定',
        showCancel: false
    });
};

/**
 * 生成缓存键
 * @param router API 路径
 * @param params 请求参数
 * @returns 缓存键
 */
export const getCacheKey = (router: string, params: Record<string, any> | null = null): string => {
    const version = uni.getAppBaseInfo?.()?.appVersionCode || '1';
    let cacheKey = `Api:${version}:${router}`;

    if (!params) {
        return cacheKey;
    }

    // 处理参数部分
    const paramValues = Object.keys(params).map(key => {
        const value = params[key];
        return typeof value === 'object'
            ? Object.values(value).join('')
            : String(value);
    }).join('');

    return cacheKey + (paramValues ? `:${paramValues}` : '');
};

/**
 * 编码 Protobuf 消息
 * @param ProtobufMessage Protobuf 消息类
 * @param params 参数对象
 * @returns 编码后的 Uint8Array
 */
export const encodeProtobuf = (ProtobufMessage: any, params: any): Uint8Array => {
    const msg = new ProtobufMessage(params);
    return ProtobufMessage.encode(msg).finish();
};

/**
 * HTTP 客户端类
 */
export class Http {
    /** Token 刷新 Promise（防止并发刷新） */
    private static refreshTokenPromise: Promise<any> | null = null;

    /**
     * 调用 API
     * @param url API 完整 URL
     * @param params 请求参数
     * @param showError 是否显示错误
     * @param errorMessage 错误提示消息
     * @param requestProtoBuf 请求 Protobuf 类
     * @param responseProtoBuf 响应 Protobuf 类
     * @param cacheTime 缓存时间（秒），0 表示不缓存
     * @returns Promise<T> 响应数据
     */
    public static callApi = <T>(
        url: string,
        params: object | null,
        showError: boolean,
        errorMessage: string,
        requestProtoBuf: any,
        responseProtoBuf: any,
        cacheTime: number = 0
    ): Promise<T> => {
        const cacheKey: string = getCacheKey(url, params as Record<string, any>);

        return new Promise<T>((resolve, reject) => {
            // 检查缓存
            if (cacheTime > 0) {
                const cacheData = CacheMgr.get(cacheKey);
                if (cacheData) {
                    try {
                        const cacheObj = JSON.parse(cacheData);
                        const cacheMsg = new responseProtoBuf(cacheObj);
                        resolve(cacheMsg as T);
                        return;
                    } catch (e) {
                        // 缓存解析失败，继续请求
                    }
                }
            }

            // 编码请求参数
            let sendParams: Uint8Array | null = null;
            if (params && typeof params === 'object' && !Array.isArray(params) && Object.keys(params).length > 0) {
                sendParams = encodeProtobuf(requestProtoBuf, params);
            }

            // 发送请求
            Http.send(url, sendParams).then((res: any) => {
                const info = responseProtoBuf.decode(res);

                // 设置缓存
                if (cacheTime > 0 && cacheKey !== '') {
                    const cacheString = JSON.stringify(info.toJSON());
                    CacheMgr.set(cacheKey, cacheString, cacheTime);
                }

                resolve(info as T);
            }).catch(async (err: any) => {
                if (showError) {
                    await showApiError(err, errorMessage);
                }
                reject(err);
            });
        });
    };

    /**
     * 使用新 Token 重试请求
     * @param url 请求 URL
     * @param uint8Array 请求数据
     * @returns Promise<any> 响应数据
     */
    private static async retryRequestWithNewToken(url: string, uint8Array: Uint8Array | null): Promise<any> {
        // 如果已经有正在进行的刷新 Token 请求，直接复用它
        if (!Http.refreshTokenPromise) {
            Http.refreshTokenPromise = Auth.refreshToken().finally(() => {
                // 请求完成后清空 Promise
                Http.refreshTokenPromise = null;
            });
        }

        // 等待 Token 刷新完成
        await Http.refreshTokenPromise;

        // 使用新 Token 重试原始请求
        return await Http.send(url, uint8Array, false);
    }

    /**
     * 发送 HTTP 请求
     * @param url 请求 URL
     * @param uint8Array 请求数据（Protobuf 编码后的数据）
     * @param allowRetry 是否允许 Token 过期后重试
     * @returns Promise<Uint8Array> 响应数据
     */
    public static send(url: string, uint8Array: Uint8Array | null = null, allowRetry: boolean = true): Promise<Uint8Array> {
        return new Promise((resolve, reject) => {
            // 等待网络就绪（处理首次打开权限问题）
            Network.waitForReady().then(() => {
                const nowTime = Math.ceil(new Date().getTime() / 1000);
                const random = Math.ceil(Math.random() * 10000);

                // 构建请求头
                const headers: Record<string, string> = {
                    'withCredentials': 'true',
                    'Content-type': 'application/x-protobuf',
                    'time': String(nowTime),
                    'random': String(random),
                    'app-id': String(Config.AppId),
                    'url': url,
                    'token': Crypto.sign(url, random, nowTime, Config.APP_SECRET),
                    'x-forwarded-proto': Config.API_URL.startsWith('https') ? 'https' : 'http'
                };

                // 添加语言设置
                const currentLang = CacheMgr.get('currentLang');
                headers['lang'] = currentLang || Config.DEFAULT_LANGUAGE || 'zh';

                // 添加短 Token
                const shortToken = Auth.getShortToken();
                if (shortToken) {
                    headers['authorization'] = shortToken;
                }

                try {
                    uni.request({
                        url: url,
                        method: 'POST',
                        responseType: 'arraybuffer',
                        dataType: 'protobuf',
                        timeout: Config.TIMEOUT,
                        data: uint8Array ? new Uint8Array(Array.from(uint8Array)).buffer : '',
                        header: headers,
                        success: (res: any) => {
                            // 处理 401 未授权
                            if (res.statusCode === 401 && allowRetry) {
                                Http.retryRequestWithNewToken(url, uint8Array).then(result => {
                                    resolve(result);
                                }).catch(error => {
                                    reject(error);
                                });
                                return;
                            }

                            // 处理非 200 状态码
                            if (res.statusCode !== 200) {
                                uni.showModal({
                                    title: '服务器错误',
                                    content: `状态码: ${res.statusCode}`,
                                    confirmText: '确定',
                                    showCancel: false
                                });
                                reject(`HTTP Error: ${res.statusCode}`);
                                return;
                            }

                            // 检查响应数据
                            if (!res.data) {
                                reject('响应数据为空');
                                return;
                            }

                            // 解码 Protobuf 响应
                            const bytes = new Uint8Array(res.data);
                            const response = Protobuf.Common.Response.decode(bytes);

                            if (response.errno === 0) {
                                resolve(response.data);
                            } else {
                                reject(response.msg);
                            }
                        },
                        fail: (err: any) => {
                            console.error('HTTP 请求失败:', url, err);
                            reject(JSON.stringify(err));
                        }
                    });
                } catch (e) {
                    console.error('HTTP 请求异常:', e);
                    reject(String(e));
                }
            });
        });
    }
}
